File: algorithm_java/Week4/boj_2178.java
package Week4;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class boj_2178 {
    static StringTokenizer st;
    static int[][] arr;

    static final int size = 4;
    static final int[] rows = { -1, 1, 0, 0 };
    static final int[] cols = { 0, 0, -1, 1 };
    static final BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));

    private static void bfs(int r, int c) {
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{r, c});

        while (!q.isEmpty()) {
            int[] tmp = q.poll();
            int curR = tmp[0];
            int curC = tmp[1];

            for (int i = 0; i < size; i++) {
                int nextR = curR + rows[i];
                int nextC = curC + cols[i];

                if (nextR < 0 || nextC < 0 || nextR >= arr.length || nextC >= arr[nextR].length)
                    continue;

                if (arr[nextR][nextC] == 0)
                    continue;

                if (arr[nextR][nextC] == 1) {
                    arr[nextR][nextC] = arr[curR][curC] + 1;
                    q.add(new int[]{ nextR, nextC });
                }
            }

        }
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        arr = new int[n][m];

        for (int i = 0; i < n; i++) {
            char[] chars = br.readLine().toCharArray();

            for (int j = 0; j < chars.length; j++)
                arr[i][j] = chars[j] - '0';

        }

        bfs(0, 0);
        System.out.println(arr[n - 1][m - 1]);
    }
}

File: algorithm/algorithm_python/Algorithm_Session/Week4/boj_2178.py
import sys
from collections import deque

input = sys.stdin.readline

n, m = map(int, input().split())

arr = []
for _ in range(n):
    tmp = list(map(int, input().rstrip()))
    arr.append(tmp)

rows = [-1, 1, 0, 0]
cols = [0, 0, -1, 1]

def bfs(r, c):
    q = deque()
    q.append((r, c))

    while q:
        curR, curC = q.popleft()
        
        for i in range(4):
            nextR = curR + rows[i]
            nextC = curC + cols[i]

            if nextR < 0 or nextC < 0 or nextR >= n or nextC >= m:
                continue

            if arr[nextR][nextC] == 1:
                arr[nextR][nextC] = arr[curR][curC] + 1
                q.append((nextR, nextC))

bfs(0, 0)
print(arr[n - 1][m - 1])